### PR TITLE
Add taint logic

### PR DESCRIPTION
**Pull request by sonasingh46**
Signed-off-by: Ashutosh Kumar <sonasingh46@gmail.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
ToDo:
Add feature gate
Add unit tests

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/controller/volume/attachdetach/util/util.go [hunk 1]
```diff
@@ -397,3 +397,14 @@ func isCSIMigrationSupportedOnNode(nodeName types.NodeName, spec *volume.Spec, v
 
 	return isMigratedOnNode, nil
 }
+
+// NodeHasTaint returns true if the node has the taint passed in the argument
+// ToDo: Add unit test
+func NodeHasTaint(taint string, nodeObject v1.Node) bool{
```
Comment (sonasingh46) [inline pkg/controller/volume/attachdetach/util/util.go]: Do we already have some util function like this in code base?

File: pkg/controller/volume/attachdetach/reconciler/reconciler.go [hunk 2]
```diff
@@ -27,17 +27,23 @@ import (
...
 	kevents "k8s.io/kubernetes/pkg/kubelet/events"
 	"k8s.io/kubernetes/pkg/util/goroutinemap/exponentialbackoff"
 	"k8s.io/kubernetes/pkg/volume/util"
 	"k8s.io/kubernetes/pkg/volume/util/operationexecutor"
 )
 
+const (
+	OutOfServiceTaint = "out-of-service"
+)
+
```
Comment (sonasingh46) [inline pkg/controller/volume/attachdetach/reconciler/reconciler.go]: Thinking, if we should move this to types? Or any proper home for this constant?

File: pkg/controller/volume/attachdetach/reconciler/reconciler.go [hunk 3]
```diff
@@ -92,6 +100,7 @@ type reconciler struct {
 	actualStateOfWorld        cache.ActualStateOfWorld
 	attacherDetacher          operationexecutor.OperationExecutor
 	nodeStatusUpdater         statusupdater.NodeStatusUpdater
+	nodeLister                corelisters.NodeLister
```
Comment (sonasingh46) [inline pkg/controller/volume/attachdetach/reconciler/reconciler.go:101]: Had to add a nodeLister to get the node object. Open to suggestions.

File: pkg/controller/volume/attachdetach/util/util.go (hunk 1 above)
Comment (sonasingh46) [reply pkg/controller/volume/attachdetach/util/util.go]: Thinking to move here 
https://github.com/kubernetes/kubernetes/blob/master/pkg/util/node/node.go
And the name could `HasTaintByKey`
wdyt?

File: pkg/controller/podgc/gc_controller.go [hunk 4]
```diff
@@ -138,13 +138,29 @@ func isPodTerminating(pod *v1.Pod) bool {
...
+
+			// Add this pod to terminatedPods list only if the following conditions are met:
+			// 1. Node is not ready.
+			// 2. Node has `out-of-service` taint with `NoExecute` effect.
+			// 3. Pod tolerated the `out-of-service` taint with `NoExecute` effect.
+			taintToMatch := &v1.Taint{
+				Key:    nodeutil.OutOfServiceTaint,
+				Effect: v1.TaintEffectNoExecute,
+			}
+			if !nodeutil.IsNodeReady(node) && hasMatchingTaint(taintToMatch, node) && isPodTolerate(taintToMatch, pod) {
```
Comment (sonasingh46) [inline pkg/controller/podgc/gc_controller.go]: What if the node has a `out-of-service` taint but the effect is NOT `NoExecute`

Comment (sonasingh46) [reply pkg/controller/podgc/gc_controller.go]: hasMatchingTaint --> Returns true if the taint matches exactly the same for `key` as well as `effect`

File: pkg/controller/podgc/gc_controller.go [hunk 5]
```diff
@@ -126,7 +126,7 @@ func isPodTerminated(pod *v1.Pod) bool {
 	return false
 }
 
-// Test Shutdown 
+// Test Shutdown
```
Comment (xing-yang) [inline pkg/controller/podgc/gc_controller.go]: Can you update the comments?

File: pkg/controller/podgc/gc_controller.go (hunk 4 above)
Comment (xing-yang) [reply pkg/controller/podgc/gc_controller.go]: "NoExecute" means evicting pods. So I think we should find the matching effect as well.

File: pkg/controller/podgc/gc_controller.go [hunk 6]
```diff
@@ -138,13 +138,29 @@ func isPodTerminating(pod *v1.Pod) bool {
...
+			node, err := gcc.nodeLister.Get(string(pod.Spec.NodeName))
+			if err != nil {
+				klog.Errorf("failed to get node:%s", err)
+				continue
+			}
+
+			// Add this pod to terminatedPods list only if the following conditions are met:
+			// 1. Node is not ready.
+			// 2. Node has `out-of-service` taint with `NoExecute` effect.
+			// 3. Pod tolerated the `out-of-service` taint with `NoExecute` effect.
```
Comment (xing-yang) [inline pkg/controller/podgc/gc_controller.go]: If Pod tolerates the taint, we should not terminate it

File: pkg/controller/volume/attachdetach/reconciler/reconciler.go (hunk 3 above)
Comment (xing-yang) [reply pkg/controller/volume/attachdetach/reconciler/reconciler.go:101]: Seems reasonable to me.  Is there code that initializes this nodeLister?

File: pkg/controller/podgc/gc_controller.go (hunk 4 above)
Comment (sonasingh46) [reply pkg/controller/podgc/gc_controller.go]: Yes. The functions `hasMatchingTaint` find the matching effect as well. 
So -- `hasMatchingTaint` returns true if the node as a taint with key `out-of-service` and effect as `NoExecute` -- for all other combination it returns false.

Do you see any action item still here?

File: pkg/controller/volume/attachdetach/reconciler/reconciler.go [hunk 7]
```diff
@@ -190,10 +192,15 @@ func (rc *reconciler) reconcile() {
...
+
+			hasOutOfServiceTaint := false
+			if utilfeature.DefaultFeatureGate.Enabled(features.NonGracefulFailover) {
+				hasOutOfServiceTaint, err = rc.hasOutOfServiceTaint(attachedVolume.NodeName)
+				if err != nil {
+					klog.Errorf("failed to get taint specs for node %s: %s", attachedVolume.NodeName, err.Error())
+				}
 			}
+
 			// TODO: Check CSIDriver SafeDetach
```
Comment (xing-yang) [inline pkg/controller/volume/attachdetach/reconciler/reconciler.go]: Can you remove this "// TODO: Check CSIDriver SafeDetach"?  This is from previous version of the KEP.

File: pkg/util/node/node.go [hunk 8]
```diff
@@ -164,3 +168,28 @@ func GetNodeIP(client clientset.Interface, name string) net.IP {
 	}
 	return nodeIP
 }
+
+// HasTaintByKey returns true if the node has the taint by the key passed in the argument.
+func HasTaintByKey(taintKey string, node v1.Node) bool {
```
Comment (xing-yang) [inline pkg/util/node/node.go]: Check here:
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.3/pkg/util/taints/taints.go#L269

File: pkg/controller/volume/attachdetach/util/util.go (hunk 1 above)
Comment (xing-yang) [reply pkg/controller/volume/attachdetach/util/util.go]: See here:
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.3/pkg/util/taints/taints.go#L269

File: pkg/controller/volume/attachdetach/reconciler/reconciler.go (hunk 2 above)
Comment (xing-yang) [reply pkg/controller/volume/attachdetach/reconciler/reconciler.go]: Move it to the following file?
https://github.com/kubernetes/kubernetes/blob/v1.24.0-alpha.3/pkg/util/taints/taints.go#L269

File: pkg/util/node/node.go (hunk 8 above)
Comment (sonasingh46) [reply pkg/util/node/node.go]: This function checks for `taint key` as well as effect. I somehow assumed that we need to check only for `key`. 
But it makes sense that, here everything on taint will be triggered only if the taint has `out-of-service` key and effect as `NoExecute`

Comment (sonasingh46) [reply pkg/util/node/node.go]: Will make the changes

**Comment by xing-yang** [comment]
LGTM

**Pull request merged**